activerecord/lib/active_record/associations/association.rb

```diff
@@ -40,7 +40,9 @@ def reset
       end
 
       # Reloads the \target and returns +self+ on success.
+      # The QueryCache is cleared.
       def reload
+        owner.class.connection.clear_query_cache if loaded?
```

Comment:
Use `klass.connection` instead of `owner.class.connection`, since both connections are not always the same.

Reply:
If this association is not loaded, `reload` on this association may be affected by the query cache, that is an edge case though.

Reply:
Indeed, I added the `if loaded?` because `reload` seems to be called the first time you call/load a singular associated record, which I'm unsure as to whether it's on purpose.

Reply:
Unless removing `if loaded?`, the following example is failed.

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")

ActiveRecord::Schema.define do
  create_table :posts

  create_table :comments do |t|
    t.integer :post_id
    t.string :body
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class ActiveRecordInconsistentCachingTest < Minitest::Test
  def connection
    ActiveRecord::Base.connection
  end

  def setup
    connection.enable_query_cache!
  end

  def teardown
    connection.disable_query_cache!
  end

  def test_reloading_has_many_associations_is_not_clearing_cache
    Post.create!
    post = Post.create!
    comment = post.comments.create!
    post.comments.to_a # add request to cache

    connection.uncached do
      comment.update!(body: "foo")
    end

    post = Post.find(post.id)

    assert_equal [comment.body], post.comments.reload.map(&:body)
  end
end
```

Reply:
Thanks! I'll update the PR shortly!